### PR TITLE
[Filter/MXNet] Add a sub-plugin filter for MXNet @open sesame 5/2 13:33

### PR DESCRIPTION
**Pull request by minty99**
**Based on PR #2831 by @kbumsik**

This PR is to newly add MXNet subplugin support for tensor_filter.

We did these things:
- Rebased the commits of PR #2831 on current main branch
- Fixed some bugs on the test code
- Add/Modify some comments into doxygen style
- Reformatted the code to follow the coding convention in `.clang-format` of the root directory

### Building MXNet
We can follow instructions on original PR #2831, but we tested on the latest version of MXNet.

### Self evaluation
1. Build test: [*]Passed [ ]Failed [ ]Skipped
2. Run test: [*]Passed [ ]Failed [ ]Skipped

### How to evaluate
Warning: The following test example will download >1.5GB of data from MXNet for dataset
```
cd nnstreamer-git
ninja -C build install
cd build/tests/nnstreamer_filter_mxnet
./download_and_run.sh    # This will run simple_test_mxnet
```

Note: I am working on this with @makesource @SungwonShim

### CODE REVIEW

File: tests/meson.build
Comment (jaeyun-jung) [inline tests/meson.build:357]: please check install directory: it would be unittest_install_dir, not unittest_tests_install_dir.

File: ext/nnstreamer/tensor_filter/tensor_filter_mxnet.cc
Comment (jaeyun-jung) [inline ext/nnstreamer/tensor_filter/tensor_filter_mxnet.cc:86]: recommend to include <nnstreamer_plugin_api.h>, not tensor_common.h.

File: ext/nnstreamer/tensor_filter/tensor_filter_mxnet.cc
Comment (jaeyun-jung) [inline ext/nnstreamer/tensor_filter/tensor_filter_mxnet.cc:57]: minor typo? spece > space

**Comment by anyj0527** [comment]
Please rebase to main branch

**Comment by minty99** [comment]
We're working on reflecting the review comments and trying to resolve some problems that can occur when running this subplugin on Android devices.

**Comment by anyj0527** [comment]
I verified the model "Inception-BN.json" used in test with simple gst-launch-1.0 pipeline. Please check.
```
$ gst-launch-1.0 filesrc location=orange.png ! pngdec !  videoscale ! videoconvert ! \
video/x-raw,format=RGB,width=224,height=224 ! tensor_converter ! \
tensor_transform mode=arithmetic option=typecast:float32,per-channel:true@0,add:-123.68@0,add:-116.779@1,add:-103.939@0 ! \
tensor_transform mode=dimchg option=0:2 ! \
tensor_filter framework=mxnet model=Inception-BN.json input=1:3:224:224 inputtype=float32 inputname=data output=1 outputtype=float32 outputname=argmax_channel custom=input_rank=4,enable_tensorrt=false latency=1 ! \
filesink location=result.log 
...

$ od -tf4 result.log
0000000             950
0000004
```
950 is the index of orange (among imagenet labels)

File: ext/nnstreamer/tensor_filter/tensor_filter_mxnet.cc
Comment (wooksong) [inline ext/nnstreamer/tensor_filter/tensor_filter_mxnet.cc:107]: Is it necessary? It looks possible to print good error messages without ```kFileUrl```

File: ext/nnstreamer/tensor_filter/tensor_filter_mxnet.cc
Comment (wooksong) [inline ext/nnstreamer/tensor_filter/tensor_filter_mxnet.cc:135]: Does including ```https://github.com/apache/incubator-mxnet/blob/master/include/mxnet/lib_api.h```  also make compile errors?

File: tests/nnstreamer_filter_mxnet/download_and_run.sh
Comment (myungjoo) [reply tests/nnstreamer_filter_mxnet/download_and_run.sh:11]: Nope. codes in main nnstreamer.git should use LGPL 2.1

@minty99 For test scripts, being LGPL 2.1 doesn't incur problems. Please apply LGPL 2.1 instead of Apache-2.0

File: ext/nnstreamer/tensor_filter/tensor_filter_mxnet.cc
Comment (minty99) [reply ext/nnstreamer/tensor_filter/tensor_filter_mxnet.cc:135]: In my opinion, the original author seems to want to support all data types NNStreamer supports defined here: https://github.com/nnstreamer/nnstreamer/blob/main/gst/nnstreamer/include/tensor_typedef.h#L153

But I think `lib_api.h` you referred is more appropriate choice and data types mxnet supports listed [here](https://mxnet.apache.org/versions/1.7.0/api/python/docs/api/ndarray/ndarray.html#mxnet.ndarray.Cast) is matched with the `lib_api.h`. Therefore I can change the TypeFlag enum to use items listed in `lib_api.h` only and remove items which are not listed there. Then some data types will be unsupported but it will be safer choice.

File: ext/nnstreamer/tensor_filter/tensor_filter_mxnet.cc
Comment (minty99) [reply ext/nnstreamer/tensor_filter/tensor_filter_mxnet.cc]: Changed like
```cpp
const static std::string kFileLocation
    = "<nnstreamer_repo_home>/ext/nnstreamer/tensor_filter/tensor_filter_mxnet.cc";
```
And replace usage of kFileUrl to use kFileLocation.

**Comment by minty99** [comment]
- I have reflected all feedbacks from code review in commits after [f7318a6](https://github.com/nnstreamer/nnstreamer/pull/3742/commits/f7318a6d85a76cf0a7fdfe57a6865a9859e151ca).
- @makesource added some fixes about Android build.
- Rebased the PR on latest main branch.

**Review by anyj0527 — APPROVED** [review]
LGTM 👍. I verified this new filter and its test. Thanks for your contribution

File: ext/nnstreamer/tensor_filter/tensor_filter_mxnet.cc
Comment (wooksong) [reply ext/nnstreamer/tensor_filter/tensor_filter_mxnet.cc]: @minty99 
If you want to specify the file where the errors occur, you can find more elegant ways (e.g., built-in macros supported by GCC or CLANG) than using the hard-coded path.

**Comment by kbumsik** [comment]
@minty99 @makesource @SungwonShim
Hi, original PR author here.

As a person who contributed quite a few random open source projects, I have had pleasantly unexpected things like this.

I remember things suddenly got quite busy when I stopped updating PR...
Thank you so much for taking over my PR and making it complete.

Excellent job 👍

File: tests/nnstreamer_filter_mxnet/Predictor.hh
Comment (jaeyun-jung) [inline tests/nnstreamer_filter_mxnet/Predictor.hh:10]: plz check the license - apache-2.0 before merging this PR. LGPL-2.1 on nnstreamer repository.

Comment (jaeyun-jung) [inline tests/nnstreamer_filter_mxnet/Predictor.hh:10]: plz check the license - apache-2.0 before merging this PR. LGPL-2.1 on nnstreamer repository.

**Comment by jaeyun-jung** [comment]
please check the license and rebase to fix conflict.

File: tests/nnstreamer_filter_mxnet/Predictor.hh
Comment (kbumsik-angelswing) [reply tests/nnstreamer_filter_mxnet/Predictor.hh:10]: Not PR author but it is already discussed here I guess: https://github.com/nnstreamer/nnstreamer/pull/3742#discussion_r861433098

Comment (kbumsik) [reply tests/nnstreamer_filter_mxnet/Predictor.hh:10]: Not PR author but it is already discuessed here I guess: https://github.com/nnstreamer/nnstreamer/pull/3742#discussion_r861433098

Comment (jaeyun-jung) [reply tests/nnstreamer_filter_mxnet/Predictor.hh:10]: Then it should be lgpl-2.1.

Comment (jaeyun-jung) [reply tests/nnstreamer_filter_mxnet/Predictor.hh:10]: 라이선스 변경이 불가능한 상황인지요?

Comment (minty99) [reply tests/nnstreamer_filter_mxnet/Predictor.hh:10]: Maybe this file is similar case to https://github.com/nnstreamer/nnstreamer/pull/3742#discussion_r861433098 (which is mentioned by @kbumsik) - there's no problem, we just missed that this file is also Apache license.

We will change it into LGPL-2.1 in next commit :) Thanks!

File: ext/nnstreamer/tensor_filter/tensor_filter_mxnet.cc
Comment (minty99) [reply ext/nnstreamer/tensor_filter/tensor_filter_mxnet.cc]: @wooksong It seems what you mentioned is `__FILE__` macro, so I changed `kFileLocation` to `__FILE__`.

File: tests/nnstreamer_filter_mxnet/Predictor.hh
Comment (minty99) [reply tests/nnstreamer_filter_mxnet/Predictor.hh]: Resolved in  https://github.com/nnstreamer/nnstreamer/pull/3742/commits/d49c44c5dbcf5627a1db2493a19806c3a7f33e67

**Comment by minty99** [comment]
- Rebased this branch on latest `main`.
- Reflected reviews from maintainers. (changed licenses of test code, use better way to generate filename in error messages)
- Fixed the CI errors (About doxygen tags in Predictor.hh)

**Review by jaeyun-jung — APPROVED** [review]

**Pull request merged**
